flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/TieredStorageNettyServiceImpl.java

```diff
@@ -0,0 +1,191 @@
+            TieredStoragePartitionId partitionId,
+            TieredStorageSubpartitionId subpartitionId,
+            BufferAvailabilityListener availabilityListener) {
+        List<NettyServiceProducer> serviceProducers = registeredServiceProducers.get(partitionId);
+        if (serviceProducers == null) {
+            return new TieredStoreResultSubpartitionView(
+                    availabilityListener, new ArrayList<>(), new ArrayList<>(), new ArrayList<>());
+        }
+        List<Queue<NettyPayload>> queues = new ArrayList<>();
+        List<NettyConnectionId> nettyConnectionIds = new ArrayList<>();
+        for (NettyServiceProducer serviceProducer : serviceProducers) {
+            LinkedBlockingQueue<NettyPayload> queue = new LinkedBlockingQueue<>();
+            NettyConnectionWriterImpl writer = new NettyConnectionWriterImpl(queue);
+            serviceProducer.connectionEstablished(subpartitionId, writer);
+            nettyConnectionIds.add(writer.getNettyConnectionId());
+            queues.add(queue);
+            registeredAvailabilityListeners.put(
+                    writer.getNettyConnectionId(), availabilityListener);
+        }
+        return new TieredStoreResultSubpartitionView(
+                availabilityListener,
+                queues,
+                nettyConnectionIds,
+                registeredServiceProducers.get(partitionId));
+    }
+
+    /**
+     * Notify the {@link ResultSubpartitionView} to send buffer.
+     *
+     * @param connectionId connection id indicates the id of connection.
+     */
+    public void notifyResultSubpartitionViewSendBuffer(NettyConnectionId connectionId) {
+        BufferAvailabilityListener listener = registeredAvailabilityListeners.get(connectionId);
+        if (listener != null) {
+            listener.notifyDataAvailable();
+        }
+    }
+
+    /**
+     * Set up input channels in {@link SingleInputGate}.
+     *
+     * @param partitionIds partition ids indicates the ids of {@link TieredResultPartition}.
+     * @param subpartitionIds subpartition ids indicates the ids of subpartition.
+     */
+    public void setUpInputChannels(
+            TieredStoragePartitionId[] partitionIds,
+            TieredStorageSubpartitionId[] subpartitionIds,
+            List<Supplier<InputChannel>> inputChannelProviders,
+            NettyConnectionReaderAvailabilityAndPriorityHelper helper) {
+        checkState(partitionIds.length == subpartitionIds.length);
+        checkState(subpartitionIds.length == inputChannelProviders.size());
+        for (int index = 0; index < partitionIds.length; ++index) {
+            TieredStoragePartitionId partitionId = partitionIds[index];
+            TieredStorageSubpartitionId subpartitionId = subpartitionIds[index];
+
+            Map<TieredStorageSubpartitionId, Integer> channelIndexes =
+                    registeredChannelIndexes.getOrDefault(partitionId, new ConcurrentHashMap<>());
```

Comment:
Maybe we can simplify it as `registeredChannelIndexes.computeIfAbsent().put()` mode.